src/core/atomic.d

```diff
@@ -35,9 +35,13 @@ private
     template HeadUnshared(T)
     {
         static if( is( T U : shared(U*) ) )
-            alias shared(U)* HeadUnshared;
+            alias HeadUnshared = shared(U)*;
+        else static if( is( T U : shared(U[]) ) )
+            alias HeadUnshared = shared(U)[];
+        else static if (is (shared T : T))
+            alias HeadUnshared = T;
         else
-            alias T HeadUnshared;
+            alias HeadUnshared = shared T;
```

Comment:
> HeadUnshared didn't remove shared from structs before your PR, so the bug description is confusing.

The problem is that HeadUnshared is never instantiated with fully shared types, only ever with tail-shared ones or completely unshared ones. That happens because atomicLoad itself strips shared already, and only then does it apply HeadUnshared. That means, the instantiations of HeadUnshared have been nops.

Now, shared needs to be re-added at some point to fix the issue with structs. I hadn't really thought this all through before, so my solution was a bit hacky.

I've now renamed HeadUnshared to TailShared and made it so that it constructs a type with a shared tail, ignoring T's given shared-ness. If T has no tail (no indirections), an unshared T is returned. If T's tail cannot be typed independently (aggregates with indirections), a shared T is returned.

Hope it makes more sense now. Tell me if this version is better. I'd squash the commits then.